app/services/renderer_service.ts

```diff
@@ -0,0 +1,142 @@
+    number,
+    {
+      matrix: Matrix
+      config: RenderConfig
+      index: number
+      gifs: Buffer[]
+      timeout?: NodeJS.Timeout
+    }
+  > = {}
+
+  async initialize() {
+    logger.info('Initializing renderer service')
+
+    const matrices = await Matrix.query()
+
+    for (const matrix of matrices) {
+      await this.addMatrix(matrix)
+    }
+  }
+
+  async addMatrix(matrix: Matrix) {
+    if (this.renders[matrix.id]) {
+      return
+    }
+
+    let config: RenderConfig = { version: 1, duration: undefined, panels: [] }
+
+    try {
+      config = await renderConfigValidator.validate(JSON.parse(matrix.config))
+    } catch (error) {
+      logger.error(error)
+    }
+
+    if (config.panels.length === 0) {
+      return
+    }
+
+    this.renders[matrix.id] = {
+      matrix,
+      config: config,
+      index: 0,
+      gifs: [],
+    }
+
+    await this.renderMatrix(matrix)
+
+    this.initializeMatrixLoop(matrix)
+  }
+
+  private initializeMatrixLoop(matrix: Matrix) {
+    if (!this.renders[matrix.id]) {
+      return
+    }
+
+    if (this.renders[matrix.id].timeout) {
+      clearTimeout(this.renders[matrix.id].timeout)
+    }
+
+    const nextIndex = (this.renders[matrix.id].index + 1) % this.renders[matrix.id].gifs.length
+    const nextDuration =
+      (this.renders[matrix.id].config.panels[this.renders[matrix.id].index].duration ||
+        this.renders[matrix.id].config.duration ||
+        10) * 1000
+
+    this.renders[matrix.id].timeout = setTimeout(() => {
+      this.renders[matrix.id].index = nextIndex
+      emitter.emit('matrix:render:updated', matrix)
+      this.initializeMatrixLoop(matrix)
+    }, nextDuration)
+  }
+
+  deleteMatrix(matrix: Matrix) {
+    if (!this.renders[matrix.id]) {
+      return
+    }
+
+    if (this.renders[matrix.id].timeout) {
+      clearTimeout(this.renders[matrix.id].timeout)
+    }
+
+    delete this.renders[matrix.id]
+  }
+
+  private async renderMatrix(matrix: Matrix) {
+    if (!this.renders[matrix.id]) {
+      return
+    }
+
+    logger.info(`Rendering matrix ${matrix.id}`)
+
+    const { config } = this.renders[matrix.id]
+
+    for (const panel of config.panels) {
+      const gif = await new Promise<Buffer | undefined>((resolve) => {
+        let filePath = app.makePath(`testFiles/pacman.gif`)
+
+        switch (panel.gifName) {
+          case 'tardis':
+            filePath = app.makePath(`testFiles/tardis.gif`)
+            break
+          case 'life':
+            filePath = app.makePath(`testFiles/life.gif`)
+            break
+          case 'homer':
+            filePath = app.makePath(`testFiles/homer.gif`)
+            break
+          case 'pacman':
+            filePath = app.makePath(`testFiles/pacman.gif`)
+            break
+        }
+
+        return fs.readFile(filePath, (err, data) => {
+          if (err) {
+            logger.error(err)
+            return
+          }
+
+          resolve(data)
+        })
+      })
```

Comment:
```suggestion
      // TODO: this can be moved to the root of the file
      const GIF_PATHS = {
        pacman: () => app.makePath(`testFiles/pacman.gif`),
        tardis: () => app.makePath(`testFiles/tardis.gif`),
        life: () => app.makePath(`testFiles/life.gif`),
        homer: () => app.makePath(`testFiles/homer.gif`),
      }
      type GifNames = keyof typeof GIF_PATHS
      function isGifName(gifName: string): gifName is GifNames {
        return gifName in GIF_PATHS
      }

      const gifName = isGifName(panel.gifName) ? panel.gifName : 'pacman'
      const filePath = GIF_PATHS[gifName]()

      let gif
      try {
        gif = await readFile(filePath)
      } catch (error) {
        logger.error(error)
      }
```

Le type, `isGifName` et `GIF_PATHS` peuvent êtres bougés dans un autre fichier ou à la racine de celui là.
La principale modif étant la suppression du new Promise pour privilégier fs/promises, le type + l'objet n'est que du "bonus"